### PR TITLE
Updated GEMS minting, staking contract, scripts, front-end

### PR DESCRIPTION
**Pull request by ablin42**
**PR includes**

**Blockchain**

- New staking contract
- Removed timestamp mapping
- Added isStaking mapping
- Added amount in struct
- Minimum Gems Amount modifier
- onlyOneStake modifier
- Staking now accepts gems  (100 min. to play)
- Unstaking rewards gems (100 for now, see comment)

**Frontend**

- Updated constants
- Updated ABIs
- Updated Play to handle gems approval
Current play behavior => Approve NFT => Approve Gems => Stake => Unstake

### CODE REVIEW

File: blockchain/contracts/Staking.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    event NFTStaked(address owner, uint256 tokenId, uint256 gemsAmount);
+    event NFTUnstaked(address owner, uint256 tokenId, uint256 rewardedAmount);
+    event Claimed(address owner);
+
+    // Map staker address to stake details
+    mapping(address => Stake) public stakes;
+    // Easier to manage than verifying the content of the Stake Struct which can be tricky due to deletion
+    mapping(address => bool) public isStaking;
```
Comment (rkhadder) [inline blockchain/contracts/Staking.sol:29]: Another way I've seen this done is having an `isInitialized` boolean on the struct that is set to true for valid structs and defaults to false if there is no reference for that address

Comment (rkhadder) [reply blockchain/contracts/Staking.sol:29]: Just sharing - up to you what to do here

File: blockchain/contracts/Staking.sol [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    // Allows only one NFT staked at a time per user
+    modifier onlyOneStake() {
+        require(!isStaking[msg.sender], "Only one stake is allowed at once");
```
Comment (rkhadder) [inline blockchain/contracts/Staking.sol:33]: nit: improve the error message here 'This address is already staking a character. Only one character per address can be staked at a time.'

File: blockchain/contracts/Staking.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        _;
+    }
+
+    modifier minimumAmount(uint256 gemsAmount) {
+        require(
+            gemsAmount >= STAKING_FEE && gemsBalance(msg.sender) >= gemsAmount,
+            "You need to have at least 100 gems to stake"
```
Comment (rkhadder) [inline blockchain/contracts/Staking.sol:40]: nit: 'Insufficient amount of gems. You need to have at least 100 gems to stake.'

**Review by rkhadder — APPROVED** [review]
LGTM - non blocking comments

File: blockchain/contracts/Gems.sol [hunk 4]
```diff
@@ -6,9 +6,7 @@ import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 
 contract Gems is ERC20, ERC20Burnable, Ownable {
-    constructor() ERC20("Gems", "GEMS") {}
-
-    function mint(address to, uint256 amount) public onlyOwner {
-        _mint(to, amount);
+    constructor() ERC20("Gems", "GEMS") {
```
[the remark this answers is not in the document]
Comment (ablin42) [reply blockchain/contracts/Gems.sol:9]: Dungeon Gems sounds good, TICKER (shorthand) should prob be DGEMS or GEMS though imo

File: blockchain/contracts/Staking.sol (hunk 1 above)
Comment (ablin42) [reply blockchain/contracts/Staking.sol:29]: Sounds definitely better, will update

File: blockchain/contracts/Staking.sol (hunk 2 above)
Comment (ablin42) [reply blockchain/contracts/Staking.sol:33]: I'll try to make it clearer in as few words as possible (more characters means more gas to pay!)

File: blockchain/contracts/Staking.sol (hunk 3 above)
Comment (ablin42) [reply blockchain/contracts/Staking.sol:40]: Agreed, i'll add the Insufficient amount:

File: blockchain/contracts/Gems.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkhadder) [reply blockchain/contracts/Gems.sol:9]: There is no such thing as ticker collision right?

[the remark this answers is not in the document]
Comment (ablin42) [reply blockchain/contracts/Gems.sol]: correct

**Pull request merged**